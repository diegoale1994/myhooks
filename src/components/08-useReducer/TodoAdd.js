import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }
        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input type="text" placeholder="Aprender ..." autoComplete="off" name="description" className="form-control" onChange={handleInputChange} value={description} />
                <button type="submit" className="btn btn-primary mt-1 btn-block">Agregar</button>
            </form>
        </>
    )
}
