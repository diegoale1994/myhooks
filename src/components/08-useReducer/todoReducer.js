export const todoReducer = (state = [], action) => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
        case 'delete':
            return state.filter(t => t.id !== action.payload);
        case 'toggle-old':
            return state.map(t => {
                if (t.id == action.payload) {
                    return { ...t, done: !t.done };
                } else {
                    return t;
                }
            })
        case 'toggle':
            return state.map(todo =>
                (todo.id === action.payload)
                    ? { ...todo, done: !todo.done }
                    : todo);
        default:
            return state;
    }

}