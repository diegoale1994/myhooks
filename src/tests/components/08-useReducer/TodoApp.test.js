import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import React from 'react';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from "../../fixtures/demoTodos";

describe('pruebas en TodoApp', () => {

    const wrapper = shallow(<TodoApp />)

    Storage.prototype.setItem = jest.fn(() => {})

    test('should mostrarse correctamente', () => {
        expect(shallow).toMatchSnapshot();
    });

    test('should agregar un todo', () => {
        const wrapper = mount(<TodoApp />);
        act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
        })
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp ( 2 )');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    })

    test('should eliminar un todo', () => {
        expect(wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0])).toBe(1);
    })
    

})
