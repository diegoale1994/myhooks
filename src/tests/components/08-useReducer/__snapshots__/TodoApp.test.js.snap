// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en TodoApp should mostrarse correctamente 1`] = `[Function]`;
